### PR TITLE
Fix race with containerd events stream on restore

### LINKED ISSUE #28687: Flaky test TestCleanupMountsAfterDaemonAndContainerKill
https://jenkins.dockerproject.org/job/Docker%20Master/7625/label=fedora23%20&&%20devicemapper/console

Daemon logs https://gist.github.com/tonistiigi/5a8edf4190efe82c7d84720ae86a05c2 show lot of things going wrong. Multiple errors/warnings in restore path and in the end daemon fails to mark the container as stopped.

https://github.com/docker/docker/pull/28590 fits with the timing but uncertain if it really is the regression point.

@mlaventure

### PR DESCRIPTION
**Pull request by mlaventure**
Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

--

Closes #28687

/cc @tonistiigi

### CODE REVIEW

File: libcontainerd/client_linux.go [hunk 1]
```diff
@@ -470,16 +463,33 @@ func (clnt *client) Restore(containerID string, attachStdio StdioCallback, optio
 				logrus.Warnf("libcontainerd: restore was called on a fully synced container (%s)", containerID)
 				return nil
 			}
-			// the container is not running so we need to fix the state within docker
-			ev = &containerd.Event{
-				Type:   StateExit,
-				Status: 1,
+		} else if ev.Pid != InitFriendlyName || ev.Type != StateExit {
```
Comment (tonistiigi) [inline libcontainerd/client_linux.go]: Shouldn't this check be in `getContainerLastEventSinceTime`? Also, what is the difference in having no last event(line 457) and having last event that is not exit. Shouldn't we wait in both cases if containerd already knows the container is gone.

Comment (tonistiigi) [inline libcontainerd/client_linux.go]: How can `getContainer` return nil. I see `appendContainer` only in start/restore paths.

Comment (mlaventure) [inline libcontainerd/client_linux.go]: duh! 

Well, that leave only the case where the event logs was rotated then, and the call to `getContainer` becomes useless.

**Comment by mlaventure** [comment]
@tonistiigi updated

File: libcontainerd/client_linux.go [hunk 2]
```diff
@@ -456,30 +451,36 @@ func (clnt *client) Restore(containerID string, attachStdio StdioCallback, optio
...
-				return nil
-			}
-			// the container is not running so we need to fix the state within docker
-			ev = &containerd.Event{
-				Type:   StateExit,
-				Status: 1,
+		if ev != nil && ev.Pid != InitFriendlyName || ev.Type != StateExit {
+			// Wait a while for the exit event
+			timeout := time.After(10 * time.Second)
+			stop := false
```
Comment (LK4D4) [inline libcontainerd/client_linux.go]: not a fan of such pattern, create helper and return from select if you don't like `break LABEL`

File: libcontainerd/client_linux.go [hunk 3]
```diff
@@ -456,30 +451,36 @@ func (clnt *client) Restore(containerID string, attachStdio StdioCallback, optio
...
-				logrus.Warnf("libcontainerd: restore was called on a fully synced container (%s)", containerID)
-				return nil
-			}
-			// the container is not running so we need to fix the state within docker
-			ev = &containerd.Event{
-				Type:   StateExit,
-				Status: 1,
+		if ev != nil && ev.Pid != InitFriendlyName || ev.Type != StateExit {
+			// Wait a while for the exit event
+			timeout := time.After(10 * time.Second)
```
Comment (LK4D4) [inline libcontainerd/client_linux.go]: This will leak timer. Better use `time.NewTimer` and Stop it.

File: libcontainerd/client_linux.go [hunk 4]
```diff
@@ -456,30 +451,36 @@ func (clnt *client) Restore(containerID string, attachStdio StdioCallback, optio
...
+				default:
+					ev, eerr = clnt.getContainerLastEvent(containerID)
+					if eerr != nil {
+						stop = true
+						break
+					}
+					if ev != nil && ev.Pid == InitFriendlyName && ev.Type == StateExit {
+						stop = true
+					} else {
+						time.Sleep(100 * time.Millisecond)
```
Comment (LK4D4) [inline libcontainerd/client_linux.go]: I'd better try to use Ticker. Don't like sleeps :)

**Review by LK4D4 — CHANGES_REQUESTED** [review]

File: libcontainerd/client_linux.go (hunk 2 above)
Comment (mlaventure) [inline libcontainerd/client_linux.go]: I had forgotten that `break LABEL` was skipping the for!

File: libcontainerd/client_linux.go [hunk 5]
```diff
@@ -456,30 +451,37 @@ func (clnt *client) Restore(containerID string, attachStdio StdioCallback, optio
...
+		if ev != nil && ev.Pid != InitFriendlyName || ev.Type != StateExit {
+			// Wait a while for the exit event
+			timeout := time.NewTimer(10 * time.Second)
+			tick := time.NewTicker(100 * time.Millisecond)
+		stop:
+			for {
+				select {
+				case <-timeout.C:
+					break stop
+				default:
```
Comment (LK4D4) [inline libcontainerd/client_linux.go]: You just called `getContainerLastEvent`, so you'll call it twice in a row. Replace `default` with `case <-tick.C`

Comment (mlaventure) [inline libcontainerd/client_linux.go]: done

Comment (LK4D4) [inline libcontainerd/client_linux.go]: not really, just replace default with case <-tick.C :)

Comment (mlaventure) [inline libcontainerd/client_linux.go]: haha, sure :)

**Review by LK4D4 — APPROVED** [review]

**Comment by LK4D4** [comment]
LGTM

**Comment by tonistiigi** [comment]
LGTM

**Pull request merged**